### DISCUSSION TITLE
Your model is not yet fully trained. Missing estimates for:

### DISCUSSION THREAD

**Discussion by sirox84** [moj-analytical-services/splink#1766] [General]
Hello everybody,

How I can fully train my model there are some variables i Cannot train.
Thanks in advance

Alberto

**Comment by ADBond** [comment]
This can happen for a couple of reasons:
* the comparisons you are using have levels that do not show up in the data (or at least not commonly enough to show up during training)
* when using [`linker.estimate_parameters_using_expectation_maximisation()`](https://moj-analytical-services.github.io/splink/linker.html#splink.linker.Linker.estimate_parameters_using_expectation_maximisation) you cannot train any variables you use in the blocking rule, so you need to run this at least twice
Further details on how to train your model can be found [in the model training tutorial](https://moj-analytical-services.github.io/splink/demos/tutorials/04_Estimating_model_parameters.html)

If you are able to share the code that you are running and the output you are seeing from Splink then it might be easier to pinpoint exactly what is happening in your case

**Reply by sirox84** [reply]
`from splink.duckdb.linker import DuckDBLinker
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl
from splink.duckdb.blocking_rule_library import block_on
from splink.datasets import splink_datasets
from splink.duckdb.duckdb_comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
)


df = pandasDF

settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name", "surname"]),
        block_on(["surname", "co_dim_tempo_nascimento"]),
        block_on(["first_name", "co_dim_tempo_nascimento"]),
    ],
    "comparisons": [
        ctl.name_comparison("first_name", term_frequency_adjustments=True),
        ctl.name_comparison("surname", term_frequency_adjustments=True),
        ctl.name_comparison("no_cidadao", term_frequency_adjustments=True),
        ctl.name_comparison("no_mae", term_frequency_adjustments=True),
        ctl.name_comparison("no_pai", term_frequency_adjustments=True),
        ctl.date_comparison("co_dim_tempo_nascimento", cast_strings_to_date=True, invalid_dates_as_null=True),
        cl.exact_match("co_dim_sexo", term_frequency_adjustments=True),
    ],
}

linker = DuckDBLinker(df, settings)

linker.estimate_probability_two_random_records_match(
    [
        "l.first_name = r.first_name and l.surname = r.surname and l.co_dim_tempo_nascimento = r.co_dim_tempo_nascimento",
        "substr(l.first_name,1,2) = substr(r.first_name,1,2) and l.surname = r.surname",
        "l.co_dim_tempo_nascimento = r.co_dim_tempo_nascimento" ,
    ],
    recall=0.6,
)


linker.estimate_u_using_random_sampling(max_pairs=1e6)

blocking_rule_for_training = block_on(["first_name", "surname"])
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)

blocking_rule_for_training = block_on("co_dim_tempo_nascimento")
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)

linker.estimate_u_using_random_sampling(max_pairs=1e6)

pairwise_predictions = linker.predict()

clusters = linker.cluster_pairwise_predictions_at_threshold(pairwise_predictions, 0.91)
panda_final= clusters.as_pandas_dataframe(limit=500000)`

**Reply by ADBond** [reply]
I can't see anything immediately wrong with this. After each training step Splink's output should give you an indication of which levels are not being trained:
* If it is `u` parameters not being trained (and you see output after `linker.estimate_u_using_random_sampling()`), you might want to increase `max_pairs` a bit - it should not take a huge amount of time if you increase this to `1e7`, or perhaps even higher depending on how long it takes. If you increase this and are still seeing levels not being trained, this probably means that the occurence of these comparison levels are too rare in your dataset, and you should probably remove them from your model
* If it is `m` parameters not being trained (which you will see in output from `linker.estimate_parameters_using_expectation_maximisation`), then again _probably_ these levels are too rare to get estimates. You _might_ be able to play around with alternate blocking rules, but probably any such estimates will not be so reliable, and it is likely best to remove these levels from your model.

If you post the output you get from Splink at these steps I can advise on what that means if what I've said isn't clear
